docs/research/USB_compliance_test.md

```diff
@@ -0,0 +1,114 @@
+
+
+
+## Device detection process debugging
+
+### USB device enumeration process
+
+USB enumeration process lets a host determine if USB device was plugged into USB
+port and what kind of device is that. Enumeration process could be divided into
+multiple stages proceeded after previous one has no error:
+
+  1. RxDetect mode (*USB 3.0 Specification Section 6.11*)
+  2. Polling LFPS (*USB 3.0 Specification Section 6.9.2*)
+  3. TSEQ Ordered Sets (*USB 3.0 Specification Section 6.4.1.1.3 and Table 6-3*)
+  4. TS1 Ordered Sets (*USB 3.0 Specification Section 6.4.1.1.3 and Table 6-4*)
+  5. TS2 Ordered Sets (*USB 3.0 Specification Section 6.4.1.1.3 and Table 6-5*)
+  6. Logical idle (sending idle data)
+  7. Exit Initialization to U0 state
+
+  >NOTE: Link Initialization and Training (stages 3,4,5) could be different
+  depending on USB 3.0 (Gen 1) or USB 3.1 (Gen 2) devices. More details are
+  available in USB 3.0 Specification Section 6.4.
+
+After last stage device should be recognized by host and communication between
+them is opened. To test device detection in system it should be checked if all
+above stages are performed correctly. Hence electrical compliance tests and link
+layer tests should be performed to compare data, timing and electrical values
+with Specification requirements.
+
+
+### Link Training and Status State Machine (LTSSM)
+
+Link Training and Status State Machine (LTSSM) is a state machine consists of 12
+link states. It is described in *USB 3.0 Specification Section 7.5*. LTSSM
+diagram is presented in figure 7.14. During USB device detection process LTSSM
+changes stages according to diagram flow. To test if device detection process is
+done correctly state machine analyze should be done.
+
+
+## Compliance tests
+
+
+### Compliance tests tools
+
+>NOTE: Official USB-IF testing software is supported under Windows only.
+
+USB Protocol Analyzer:
+
+- [Ellisys USB Explorer 280](https://www.ellisys.com/products/usbex280/index.php)
+- [Teledyne LeCroy Voyager M3x](https://teledynelecroy.com/protocolanalyzer/usb/voyager-m3x)
+- [Beagle USB 5000 v2 SuperSpeed Protocol Analyzer](https://www.totalphase.com/products/beagle-usb5000-v2-ultimate/)
+
+USB Protocol Analyzer Software:
+
+- [Totalphase Data Center Software](https://www.totalphase.com/products/data-center/)
+- [Virtual USB Analyzer](http://vusb-analyzer.sourceforge.net/)
+
+
+### Electrical compliance tests
+
+These tests verify that electrical signals send between host and device meet USB 3.0
+criteria which are described in USB 3.0 Specification Document. All details
+about tests are described in [Electrical Compliance Test Specification SuperSpeed Universal Serial Bus](https://www.usb.org/sites/default/files/SuperSpeedPHYComplianceTest_Spec1_0a.pdf).
+
+Should be considered to carry out those tests:
+
+- TD.1.1 Low Frequency Periodic Signaling TX Test
+
+- TD.1.2 Low Frequency Periodic Signaling RX Test
+
+- TD.1.3 Transmitted Eye Test
```

Comment:
@piotr-kleins @lwcislo this is what we have to consider in light of hardware capabilities that we have. Definitely going with expensive and proprietary LeCroy equipment is not our way of doing things.